File: backend/userservice/endpoint/create-profile.go
package UserEndpoint

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"services/photoservice"
	"services/userservice"
)

type contextKey string

const (
	UIDKey   contextKey = "uid"
	EmailKey contextKey = "email"
)

func CreateProfileHandler(svc *userservice.Service, photoSvc *photoservice.Service) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
			return
		}

		uid, ok := r.Context().Value(UIDKey).(string)
		if !ok || uid == "" {
			http.Error(w, "Unauthorized", http.StatusUnauthorized)
			return
		}
		email, ok := r.Context().Value(EmailKey).(string)
		if !ok || email == "" {
			log.Println("Warning: Email not found in context for user:", uid)
			email = ""
		}

		err := r.ParseMultipartForm(10 << 20)
		if err != nil {
			log.Printf("Error parsing multipart form: %v", err)
			http.Error(w, "Bad Request", http.StatusBadRequest)
			return
		}

		profileData := userservice.UserProfileData{
			FirstName: r.FormValue("firstName"),
			LastName:  r.FormValue("lastName"),
			DOB:       r.FormValue("dob"),
			Username:  r.FormValue("username"),
			Email:     email,
		}

		var photoURL string
		file, handler, err := r.FormFile("file")
		if err != nil {
			if err != http.ErrMissingFile {
				log.Printf("Error retrieving file from form: %v", err)
				http.Error(w, "Failed to retrieve file", http.StatusBadRequest)
				return
			}
			log.Printf("No file provided for profile photo. Proceeding without one.")
		} else {
			defer file.Close()
			log.Printf("Found file: %s", handler.Filename)

			photoData, uploadErr := photoSvc.UploadAndStorePhoto(context.Background(), uid, file, handler.Filename, handler.Header.Get("Content-Type"))
			if uploadErr != nil {
				log.Printf("Error uploading photo: %v", uploadErr)
				http.Error(w, "Failed to upload photo", http.StatusInternalServerError)
				return
			}

			photoURL = photoData.URL
			log.Printf("Photo successfully uploaded. URL: %s", photoURL)
		}

		profileData.PhotoURL = photoURL

		createdProfile, err := svc.CreateUserProfile(r.Context(), uid, profileData)
		if err != nil {
			log.Printf("Error creating user profile: %v", err)
			if err.Error() == "Username already taken" {
				http.Error(w, "Username is already taken. Please choose a different one.", http.StatusConflict)
			} else if err.Error() == "User profile already exists" {
				http.Error(w, "A profile for this user already exists.", http.StatusConflict)
			} else {
				http.Error(w, fmt.Sprintf("Failed to create user profile: %v", err), http.StatusInternalServerError)
			}
			return
		}

		w.Header().Set("Content-Type", "application/json")
		if err := json.NewEncoder(w).Encode(createdProfile); err != nil {
			log.Printf("Error encoding response: %v", err)
			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
			return
		}
	}
}

File: web/pages/api/user/create-profile.ts
import type { NextApiRequest, NextApiResponse } from "next";
import { getAuth } from "firebase-admin/auth";
import { app } from "@/lib/firebase/firebaseAdmin";
import { createUserProfile as createProfileService } from '@/lib/firebase/adminUtils/userService';
import { corsInstance, runMiddleware } from "@/lib/authorization/helper";
import { Formidable } from 'formidable';
import fs from 'fs';
import { uploadAndStorePhoto } from "@/lib/firebase/adminUtils/photoService";

export const config = {
    api: {
        bodyParser: false,
    },
};

const cors = corsInstance(['POST', 'OPTIONS']);

// Creating profile with authorization.
export default async function handler (req: NextApiRequest, res: NextApiResponse){
    await runMiddleware(req, res, cors);
    if(req.method === 'OPTIONS') {
        return res.status(200).end();
    }

    if(req.method !== 'POST') {
        return res.status(405).json({ message: "Method Not Allowed." });
    }

    try {
        const idToken = req.headers.authorization?.split('Bearer ')[1];
        if(!idToken) {
            return res.status(401).json({ error: "Unauthorized: No token provided." });
        }
        
        let decodedToken;
        try {
            decodedToken = await getAuth(app).verifyIdToken(idToken);
        } catch (err) {
            console.error("Error verifying ID token: ", err);
            return res.status(401).json({ error: "Unauthorized: Invalid token." });
        }

        const uid = decodedToken.uid;
        const email = decodedToken.email;
        console.log("autorized!");
        
        const formidable = new Formidable();
        const { fields, files } = await new Promise<{ fields: any, files: any }>((resolve, reject) => {
            formidable.parse(req, (err, fields, files) => {
                if (err) {
                    return reject(err);
                }
                resolve({ fields, files });
            });
        });

        const firstName = fields.firstName?.[0];
        const lastName = fields.lastName?.[0];
        const dob = fields.dob?.[0];
        const username = fields.username?.[0];

        if (!firstName || !lastName || !dob || !username) {
            return res.status(400).json({ error: 'Missing required profile data.' });
        }

        const file = files.file?.[0];
        console.log('file: ' + file);
        let photoData = null;
        
        if (file) {
            try {

                const fileContent = fs.readFileSync(file.filepath);
                console.log('File Content: ', fileContent);
                console.log('at filepath: ' + file.filepath);
                photoData = await uploadAndStorePhoto(uid, fileContent, file.originalFilename || 'unnamed-file', file.mimetype || 'application/octet-stream');
                console.log('Photo data: ' + photoData);

            } catch (uploadErr) {
                console.error("Error during photo upload: ", uploadErr);
                // Return an error without creating the profile if the photo upload fails
                return res.status(500).json({ error: "Failed to upload profile photo." });
            }
        }
        
        const newProfile = await createProfileService(uid, {
            firstName,
            lastName,
            dob,
            username,
            email,
            photoURL: photoData?.url || null, // Pass the photoURL if it exists
        });
        
        // Return both the profile and photo data in the response
        return res.status(200).json({ success: true, profile: newProfile, photoData });
    } catch (err: any) {
        console.error("An error occurred while creating user profile via API route: " + err);

        // Map specific errors to appropriate HTTP responses
        if (err.message.includes("Username already taken")) {
            return res.status(409).json({ error: "Username is already taken. Please choose a different one." });
        }
        if (err.message.includes("User profile already exists")) {
            return res.status(409).json({ error: "A profile for this user already exists." });
        }
        return res.status(500).json({ error: "Failed to create user profile due to a server error." });
    }
}
